jib-core/src/main/java/com/google/cloud/tools/jib/http/Connection.java

```diff
@@ -114,87 +109,157 @@ private static void addProxyCredentials(ApacheHttpTransport transport, String pr
             new UsernamePasswordCredentials(proxyUser, proxyPassword));
   }
 
-  private HttpRequestFactory requestFactory;
-
-  @Nullable private HttpResponse httpResponse;
+  private final boolean enableHttpAndInsecureFailover;
+  private final boolean sendAuthorizationOverHttp;
+  private final Consumer<LogEvent> logger;
+  private final Supplier<HttpTransport> httpTransportFactory;
+  private final Supplier<HttpTransport> insecureHttpTransportFactory;
 
-  /** The URL to send the request to. */
-  private final GenericUrl url;
-
-  /**
-   * Make sure to wrap with a try-with-resource to ensure that the connection is closed after usage.
-   *
-   * @param url the url to send the request to
-   */
-  @VisibleForTesting
-  Connection(URL url, HttpTransport transport) {
-    this.url = new GenericUrl(url);
-    requestFactory = transport.createRequestFactory();
+  public Connection(
+      boolean enableHttpAndInsecureFailover,
+      boolean sendAuthorizationOverHttp,
+      Consumer<LogEvent> logger) {
+    this(
+        enableHttpAndInsecureFailover,
+        sendAuthorizationOverHttp,
+        logger,
+        Connection::getHttpTransport,
+        Connection::getInsecureHttpTransport);
   }
 
-  @Override
-  public void close() throws IOException {
```

Comment:
We should still have a `close()` or `shutDown()` to actually release resources (imagine the jib-core case).

Reply:
So that means this class should maintain a list of `Response` objects that are generated. It certainly is a nice feature as a safety net, but I think we should add that later considering the value and the new code to maintain the list. I'll file a new issue.

(Honestly, I feel comfortable making the user responsible for closing `Response`, which is the way Apache HttpClient works. I've seen IDEs warn if `AutoCloseable` is left open. However, it is true a method like `shutDown()` will make all `Response`s disconnect immediately.)

Reply:
Hmm, well `HttpTransport` does have a `shutdown()` method, and it's overridden by `ApacheHttpTransport` which seems to do actually something.

Reply:
Ah, it's probably good to call `shutdown()` on `HttpTransport`. I'll make a note in #2111. At least this PR retains the current state of not calling `shutdown()` and doesn't make anything worse about this matter.